File: node/src/retry.ts
/**
 * Retry policy for transient failures.
 *
 * Retries happen on:
 *   - Network errors (DNS, connect, read timeout, TLS)
 *   - HTTP 429 (rate limit)
 *   - HTTP 500, 502, 503, 504 (transient server issues)
 *
 * Retries do NOT happen on:
 *   - 4xx other than 429 (caller error, retrying won't help)
 *   - Non-idempotent methods by default (POST, PATCH): mutations are
 *     never auto-retried. Only GET/HEAD/OPTIONS/PUT/DELETE are retried.
 *     This prevents duplicate webhook creation and duplicate retry calls.
 *
 * The `Retry-After` header wins when present and non-negative. Otherwise
 * exponential backoff with full jitter, capped at `maxDelay`.
 */

export const DEFAULT_MAX_RETRIES = 3;
export const DEFAULT_INITIAL_DELAY = 0.5; // seconds
export const DEFAULT_MAX_DELAY = 30.0; // seconds — matches Python reference
export const DEFAULT_BACKOFF_FACTOR = 2.0;

export const RETRY_STATUSES: ReadonlySet<number> = new Set([429, 500, 502, 503, 504]);

/** HTTP methods considered safe to auto-retry (idempotent). */
export const IDEMPOTENT_METHODS: ReadonlySet<string> = new Set([
  "GET",
  "HEAD",
  "OPTIONS",
  "PUT",
  "DELETE",
]);

export interface RetryPolicyOptions {
  maxRetries?: number;
  initialDelay?: number;
  maxDelay?: number;
  backoffFactor?: number;
  /**
   * When true, retry POST/PATCH too. Default false — the SDK never
   * auto-retries mutations unless the caller opts in explicitly.
   */
  retryNonIdempotent?: boolean;
}

/**
 * Configuration for automatic retries.
 *
 * Immutable — construct a new instance to tweak. Matches the Python
 * `RetryPolicy` contract:
 *   - `maxRetries`: total attempts is at most `maxRetries + 1`. 0 disables retries.
 *   - `initialDelay`: seconds before the first retry (pre-jitter).
 *   - `maxDelay`: cap on any single sleep in seconds.
 *   - `backoffFactor`: multiplier applied to delay each attempt.
 */
export class RetryPolicy {
  readonly maxRetries: number;
  readonly initialDelay: number;
  readonly maxDelay: number;
  readonly backoffFactor: number;
  readonly retryNonIdempotent: boolean;

  constructor(options: RetryPolicyOptions = {}) {
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.initialDelay = options.initialDelay ?? DEFAULT_INITIAL_DELAY;
    this.maxDelay = options.maxDelay ?? DEFAULT_MAX_DELAY;
    this.backoffFactor = options.backoffFactor ?? DEFAULT_BACKOFF_FACTOR;
    this.retryNonIdempotent = options.retryNonIdempotent ?? false;
  }

  /**
   * Decide whether to retry given attempt index, HTTP status, and method.
   *
   * `status` is undefined when the failure was a transport error before
   * the server returned a status. Transport errors are retried for any
   * method (the request never hit the server, so the "don't duplicate
   * mutations" concern doesn't apply).
   */
  shouldRetry(attempt: number, options: { status?: number; method?: string }): boolean {
    if (attempt >= this.maxRetries) return false;
    const method = (options.method ?? "GET").toUpperCase();

    if (options.status === undefined) {
      // Transport error — safe to retry even for POST/PATCH, because the
      // request never reached the server.
      return true;
    }

    if (!RETRY_STATUSES.has(options.status)) return false;

    // Server returned a retryable status. Only retry idempotent methods
    // unless the caller opted in.
    if (this.retryNonIdempotent) return true;
    return IDEMPOTENT_METHODS.has(method);
  }

  /**
   * Seconds to wait before retry `attempt` (0-indexed).
   *
   * `Retry-After` wins unambiguously when present and non-negative: the
   * server is telling us exactly how long to wait. Otherwise we use
   * exponential backoff with full jitter:
   *
   *     delay = random.uniform(0, min(maxDelay, initial * factor^attempt))
   *
   * Full jitter beats "equal jitter" and "decorrelated jitter" for
   * preventing thundering-herd recovery spikes.
   */
  computeDelay(attempt: number, options: { retryAfter?: number }): number {
    if (options.retryAfter !== undefined && options.retryAfter >= 0) {
      return Math.min(options.retryAfter, this.maxDelay);
    }
    const base = this.initialDelay * Math.pow(this.backoffFactor, attempt);
    const capped = Math.min(base, this.maxDelay);
    return Math.random() * capped;
  }
}

/**
 * Parse the `Retry-After` header into seconds.
 *
 * RFC 9110 allows two forms:
 *   - A non-negative integer (delta-seconds): `Retry-After: 120`
 *   - An HTTP-date: `Retry-After: Wed, 21 Oct 2025 07:28:00 GMT`
 *
 * Unparseable input returns `undefined` so the caller can fall back to
 * its own backoff. HTTP-date values in the past clamp to `0`.
 */
export function parseRetryAfter(header: string | null | undefined): number | undefined {
  if (header === null || header === undefined) return undefined;
  const trimmed = header.trim();
  if (!trimmed) return undefined;

  // Delta-seconds form: pure integer (possibly negative).
  if (/^-?\d+$/.test(trimmed)) {
    const n = parseInt(trimmed, 10);
    return Math.max(0, n);
  }

  // HTTP-date form.
  const parsed = Date.parse(trimmed);
  if (Number.isNaN(parsed)) return undefined;
  const nowMs = Date.now();
  const delta = (parsed - nowMs) / 1000;
  return Math.max(0, delta);
}

/** Promise-based sleep — used by the client's retry loop. */
export function sleep(seconds: number): Promise<void> {
  if (seconds <= 0) return Promise.resolve();
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

File: go/retry.go
package legalize

import (
	"math"
	"math/rand"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// Retry defaults. Match the Python SDK byte-for-byte.
const (
	DefaultMaxRetries    = 3
	DefaultInitialDelay  = 500 * time.Millisecond
	DefaultMaxDelay      = 30 * time.Second
	DefaultBackoffFactor = 2.0
)

// retryStatuses is the set of HTTP status codes that trigger a retry
// when returned by the server. Mirrors Python's RETRY_STATUSES.
var retryStatuses = map[int]struct{}{
	408: {},
	429: {},
	500: {},
	502: {},
	503: {},
	504: {},
}

// idempotentMethods is the set of HTTP methods that the SDK retries
// on transient failure by default. POST and PATCH are excluded to
// avoid duplicate webhook creation and duplicate retry-delivery
// calls. Callers can still opt into retrying those via the RequestOption
// mechanism.
var idempotentMethods = map[string]struct{}{
	http.MethodGet:     {},
	http.MethodHead:    {},
	http.MethodOptions: {},
	http.MethodPut:     {},
	http.MethodDelete:  {},
}

// RetryPolicy describes the SDK's transient-failure retry behaviour.
// Zero-value retries.MaxRetries disables retries entirely. Leave
// fields at their zero value to pick up the compiled-in defaults via
// the NewRetryPolicy helper.
type RetryPolicy struct {
	// MaxRetries caps the number of additional attempts after a
	// failure. Total requests is at most MaxRetries + 1.
	MaxRetries int
	// InitialDelay is the baseline backoff before the first retry.
	InitialDelay time.Duration
	// MaxDelay caps any single retry delay.
	MaxDelay time.Duration
	// BackoffFactor multiplies the delay each attempt (exponential).
	BackoffFactor float64
	// rng is an injectable jitter source for deterministic tests.
	// Left nil in the public API so users never see it.
	rng *rand.Rand
}

// NewRetryPolicy returns a RetryPolicy with the SDK defaults applied.
func NewRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxRetries:    DefaultMaxRetries,
		InitialDelay:  DefaultInitialDelay,
		MaxDelay:      DefaultMaxDelay,
		BackoffFactor: DefaultBackoffFactor,
	}
}

// withDefaults fills any zero fields on the policy with the compiled-in
// defaults so callers can pass a partially-populated struct.
func (p RetryPolicy) withDefaults() RetryPolicy {
	if p.InitialDelay == 0 {
		p.InitialDelay = DefaultInitialDelay
	}
	if p.MaxDelay == 0 {
		p.MaxDelay = DefaultMaxDelay
	}
	if p.BackoffFactor == 0 {
		p.BackoffFactor = DefaultBackoffFactor
	}
	return p
}

// ShouldRetry decides whether the client should issue another attempt
// after the given 0-indexed failed attempt and HTTP status code.
// Pass status = -1 for a transport-layer failure (no response).
func (p RetryPolicy) ShouldRetry(attempt int, status int) bool {
	if attempt >= p.MaxRetries {
		return false
	}
	if status < 0 {
		return true
	}
	_, ok := retryStatuses[status]
	return ok
}

// ComputeDelay returns the sleep duration before the given retry.
// The server-provided Retry-After header wins unambiguously when
// present; otherwise we apply exponential backoff with full jitter,
// capped at MaxDelay.
func (p RetryPolicy) ComputeDelay(attempt int, retryAfter *time.Duration) time.Duration {
	p = p.withDefaults()
	if retryAfter != nil && *retryAfter >= 0 {
		if *retryAfter > p.MaxDelay {
			return p.MaxDelay
		}
		return *retryAfter
	}
	base := float64(p.InitialDelay) * math.Pow(p.BackoffFactor, float64(attempt))
	cap := float64(p.MaxDelay)
	if base > cap {
		base = cap
	}
	// Full jitter: uniform random in [0, base].
	var frac float64
	if p.rng != nil {
		frac = p.rng.Float64()
	} else {
		frac = rand.Float64() //nolint:gosec // jitter, not crypto
	}
	return time.Duration(frac * base)
}

// ParseRetryAfter parses an HTTP Retry-After header to a duration.
//
// RFC 9110 allows two forms: a non-negative integer delta-seconds,
// or an HTTP-date. Both are accepted. Unparseable input returns nil
// so the caller can fall back to its own backoff policy. Past
// HTTP-date values clamp to 0.
func ParseRetryAfter(header string) *time.Duration {
	return parseRetryAfter(header)
}

// parseRetryAfter is the internal lowercase twin of ParseRetryAfter.
func parseRetryAfter(header string) *time.Duration {
	header = strings.TrimSpace(header)
	if header == "" {
		return nil
	}
	if n, err := strconv.Atoi(header); err == nil {
		if n < 0 {
			n = 0
		}
		d := time.Duration(n) * time.Second
		return &d
	}
	if t, err := http.ParseTime(header); err == nil {
		delta := time.Until(t)
		if delta < 0 {
			delta = 0
		}
		return &delta
	}
	return nil
}
